---
title: >
  Basic Protocol 1: Exploratory Data Analysis with pcaExplorer
author:
- name: Annekathrin Ludt
  affiliation: 
  - &id1 Institute of Medical Biostatistics, Epidemiology and Informatics (IMBEI), Mainz
  email: anneludt@uni-mainz.de
- name: Arsenij Ustjanzew
  affiliation: 
  - &id1 Institute of Medical Biostatistics, Epidemiology and Informatics (IMBEI), Mainz
  email: arsenij.ustjanzew@uni-mainz.de
- name: Chung Shing Rex Ha
  affiliation: 
  - *id1
  email: rexha@uni-mainz.de
- name: Harald Binder
  affiliation:
  - Institute of Medical Biometry and Statistics (IMBI), Faculty of Medicine and Medical Center, University of Freiburg 
  email: binder@imbi.uni-freiburg.de 
- name: Konstantin Strauch
  affiliation: 
  - *id1
  email: strauch@uni-mainz.de
- name: Federico Marini
  affiliation: 
  - *id1
  - &id2 Center for Thrombosis and Hemostasis (CTH), Mainz;<br>
  email: marinif@uni-mainz.de
date: "`r BiocStyle::doc_date()`"
package: "`r BiocStyle::pkg_ver('ideal')`"
output: 
  bookdown::html_document2:
  # BiocStyle::html_document:
    toc: true
    toc_float: true
    theme: cosmo
    code_folding: show
    code_download: true
editor_options: 
  chunk_output_type: console
bibliography: cpb_refs.bib
link-citations: true
---

**Compiled date**: `r Sys.Date()`

**Last edited**: `r Sys.Date()`

```{r setup, include = FALSE, cache = FALSE, eval = TRUE, echo = FALSE}
library(knitr)
opts_chunk$set(
  fig.align = "center",
  fig.show = "asis",
  eval = TRUE,
  fig.width = 10,
  fig.height = 7,
  tidy = FALSE,
  message = FALSE,
  warning = FALSE,
  size = "small",
  comment = "##",
  echo = TRUE,
  results = "markup"
)
options(replace.assign = TRUE, width = 100)
```

# Loading required packages

```{r loadLibraries, results= "hide"}
library("DESeq2")
library("topGO")
library("org.Hs.eg.db")
library("pcaExplorer")
library("ideal")
library("GeneTonic")
```


# Introductory Paragraph

pcaExplorer is a Bioconductor package [@Marini2019] which can be used as a general-purpose interactive companion tool for RNA-seq analyses, developed in the Shiny framework [@shiny2021]. 
pcaExplorer is designed to guide the user in exploring the Principal Components (PC) [@Jolliffe2002] of the data under inspection. 
Besides the Principal Component Analysis (PCA), pcaExplorer also provides tools to detect outlier samples and inspect their impact, identify genes that show particular patterns of interest, and additionally provides a functional interpretation of the principal components for further quality assessment and hypothesis generation on the input data. 

In this protocol, we describe how to launch an instance of pcaExplorer working with the data of the macrophage dataset [@Alasoo2018], which is also distributed as a Bioconductor package [@Huber2015]. 

# Necessary Resources

*Hardware*

  See detailed description in Support Protocol 1
  
*Software*

  See detailed description in Support Protocol 1
  
*Files* (maybe we can describe the sample input here and how to download from Github, I've seen that in some papers rather than describing the general input format)

**TODO: finalize input. Currently the pictures are the example data from the pcaExplorer vignette, but it would probably be better to use screenshots of the macrophage dataset? Or we describe everything in the Support Protocol? **

  pcaExplorer mainly requires 3 input files in text format. 
  The files are expected to be tab-separated, but also comma- or semicolon-separated files are accepted (see **Alternative protocol 1**). 
  The first input file is the count matrix which stores the number of times (i.e. counts) a certain feature (e.g gene) is found in each sample, as a proxy of its expression. 
  In the count matrix, the samples are stored in the columns, while the rows store the individual features (see Fig. X).
  
  ![**Figure ** Example of a count matrix - The figure shows the example of a count matrix, with the individual features (here genes) in the rows and the individual samples in the columns. The matrix represents how many times each feature was found in each sample. The shown count matrix is comma-separated, but also semicolon- and tab-separated inputs are accepted (see **Alternative protocol 1**)](Figures_Basic_1/Figure 1.png)

  The second input of pcaExplorer is the metadata file, which stores for each sample the relevant experimental variables. 
  The individual samples represent the rows of the file while the columns save the different experimental variables (see Fig. X).
  
  ![**Figure ** Example of a Metadata file - The figure shows the example of a Metadata file. The rows of the file contain the individual samples while the columns represent relevant experimental variables. The variable can change depending on the data set and research question.](Figures_Basic_1/Figure 2.png)
  Lastly, the third input (the annotation file) of pcaExplorer is optional, but highly recommended for easier interpretation of the results. 
  The annotation file contains the feature identifiers of the count matrix in the rows and at least one column called gene_name which contains a more human readable form of the feature ids (e.g. HGNC gene names [@Tweedie2021] if the features are gene ids). 
  Fig. X shows an example of an annotation file.
  
  ![**Figure ** Example of an annotation file - The annotation file contains a row for each of the features of the count matrix. Furthermore, the file contains at least one column called gene names which contains a more human readable for of the feature identifiers of the rows. In the shown example the feature ids are ENSEMBL ids while the column gene names contains HGNC gene names. ](Figures_Basic_1/Figure 3.png)

# Protocol steps

*Exploring the data with pcaExplorer* 

Before we start with the exploration of the data, the necessary packages and dependencies need to be installed and loaded. 
**Support Protocol 1** describes how to install and load the packages. 

1. Prepare the input data for pcaExplorer.
The main pcaExplorer function requires the count matrix, the metadata, and annotation to be either loaded into the environment before the main call, or offers the possibility to load these objects during runtime (which can be common for a first execution).
If loading the input files at runtime, the application can be launched with the simple command, leaving all other parameters unspecified.

```{r, eval=FALSE}
pcaExplorer()
```

If the necessary objects are already loaded in R (e.g. with any function that reads tabular text input), users can launch the Shiny application with the command:

```{r, eval=FALSE}
pcaExplorer(countmatrix = countmatrix, 
            coldata = metadata, 
            annotation = annotation)
```

-in this call, `countmatrix`, `metadata` and `annotation` have to be substituted by the names of the respective objects.
To launch the application, enter the command into the console of RStudio and press the Enter key. 
This should launch a new window with the pcaExplorer application. 
In this application you should see the Data Upload panel as shown in Fig. X.


![**Figure ** Data Upload panel - The figure shows the Data Upload panel which is the first panel you see upon launching pcaExplorer. The panel provides previews on the input data, as well as buttons to interact with the application.](Figures_Basic_1/Figure 4.png)

2. Click the 'Generate the dds and dst object' button (Fig X). 
The dds object represents a `DESeqDataSet` dataset object, while the dst is a `DESeqTransform` object [@Love2014]. 
Both objects are needed for the exploration of the data, and can be readily derived from the provided files/objects - alternatively, users can pass directly the dds and the dst objects, in case they are already loaded in the R environment (e.g. while running an analysis from the command line).
This is possible with the lines of code reported here:

```{r, eval=FALSE}
pcaExplorer(dds = my_dds,
            dst = my_dst)
```


After the generation of the two objects, the 'Select one of the following transformations for your data:' option with three blue colored buttons underneath should appear in the panel. 
Click the 'Compute variance stabilized transformed data from the `dds` object' button on the left. 

     *pcaExplorer provides three different option for data transformation. 
     The first option is to compute a variance stabilized transformed version of the data. 
     The second is a regularized logarithm transformation of the data. 
     A log2 data transformation is also offered. 
     Users are advised to choose the data transformation according to their input data and evaluate the effects of the different transformations on their data.*

3. Preview the provided input data.
Scroll to the bottom of the Data Upload panel, where a preview of the input data is provided.
You can access each of them by clicking on the green colored buttons (see Fig. X) - each of these opens up a separate modal window.

4. Explore in detail the provided counts table.
Navigate to the Counts Table panel, by clicking on the name in the tabs list at the top of the main body of the application (Fig. X). 
In this panel, the information of the count matrix is shown in a table. 
A drop-down menu at the beginning of the panel provides the possibility to change the displayed expression table (including raw counts, normalized counts, regularized logarithm transformed counts, etc.). 
Users can download these different versions of the original counts table through clicking on the green download button below (Fig. X).

![**Figure ** Counts Table panel - The figure shows the Counts Table panel. This panel provides an overview of the input counts matrix.](Figures_Basic_1/Figure 5.png)

5. Explore the correlations between samples.
Scroll down until you see the 'Sample to sample scatter plots' heading (Fig. X). 
Choose 'pearson' as correlation method from the 'Correlation method palette'. 
If you want to reduce the computation and rendering time, check that both options 'Use log2 values for plot axes and values' and 'Use a subset of max 1000 genes (quicker to plot)' are selected.
Click on the 'Run' button to generate the scatter plots (Fig. X and X.). 
  
    *Upon clicking the 'Run' button, pcaExplorer will generate a scatter plot (Fig. X) as well as a heatmap (Fig. X). 
    The scatter plot will represent the sample to sample correlation as scatter plots (lower triangular matrix of the result). 
    The precise values of the correlations between two samples can be found in the diagonal matching (??) field. 
    The heatmap (Fig. X) shows the sample to sample similarity. 
    The color indicates the similarity with dark blue indicating  a low similarity and dark red indicating identical samples (see diagonal of the heatmap).*

 **TODO: Maybe merge these pictures into one?**

![**Figure ** Sample to Sample scatter plot- The plot visualizes the sample to sample correlation as scatter plots as well as the precise values of each correlation.](Figures_Basic_1/Figure 6.png)

![**Figure ** Sample to Sample similarity heatmap - The figure shows the sample to sample similarity. Identical samples are colored in dark red, while dark blue indicates a low similarity of the samples.](Figures_Basic_1/Figure 7.png)

**Create an overview and compute some summaries on the provided data.**

6. Explore the relationships between the samples.
Navigate to the Data Overview panel (Fig. X) as explained above (**TODO: Mention exact step**). 
The panel shows the input metadata as a table in the upper section of the user interface, with some interactive features to edit its appearance or to search its content (Fig. X). 
Below the metadata table, a Sample to Sample distance heatmap can be found, representing the distances between the individual samples. 
The 'Select the distance method to use' option enables you to change the underlying distance method used for the computation.
Furthermore, you have the option to download the heatmap via the 'Download Plot' button in the lower right corner below the heatmap. 
In the field 'Save as...' you can specify the name of the file to save the heatmap to (Fig X) - this functionality is provided for any graphical output presented in pcaExplorer.


![**Figure X** Sample to Sample distance heatmap - The figure shows the sample to sample distance heatmap. Dark blue (diagonal) indicates samples with a low distance, while dark red colored samples have the highest distance. ](Figures_Basic_1/Figure 8.png)

7. Decorate the plots by encoding information on the experimental covariates.
Below the heatmap, this panel provides some further information about the input data (Fig. X & X). 
Next, color the sample by their originating cell line. 
To color the samples by this characteristic, click on the 'Group/color by' field in the grey side bar at the far left of the screen. 
Scroll down in the drop-down menu which opens upon clicking until you find the 'line_id' option - or leverage the autocompletion functionality and start typing the initials. 
Once any option is selected, scroll down to the 'Number of million of reads per sample' plot. 
This plot should have changed from an all grey plot to a colored plot (see Fig. X). 

    *In the provided sample data, the line_id parameter indicates the cell line of the respective sample. 
    The Number of million reads per sample plot shown in Fig. X can be used to evaluate the uniformity of reads in each sample. 
    Samples with a surprisingly high or low number of reads could later be detected as outliers, because of several different factors associated with the number of reads in a sample. 
    An inspection of the distribution of the overall number of reads in the individual sample can point users to the first source of differences in the samples found in subsequent analyses of the data.*

![**Figure X** Number of million reads per sample plot- The plot visualizes the number of reads per sample. The individual bars are colored by their cell line origin. ](Figures_Basic_1/Figure 9.png)

8. Change the experimental covariates of interest.
Scroll up again until you see the 'Group/color by' option. 
Delete the selected option by clicking on the white rectangle and hitting the delete key on your keyboard until the field is empty again. 
From the drop-down menu of options, select 'condition' as next parameter to color by. 
Instead of selecting 'condition' from the drop-down menu directly you can also select the option by typing it into the white rectangle. 
This widget controls the appearance of many plots in pcaExplorer, and is used to encode the grouping information (or their combinations).

  *If you inspect the Number of million of reads per sample plot again, you should observe that the coloring changed compared to the version showed in Fig. X. The 'Group/color by' option is very helpful in exploring different characteristics of the data and providing context to individual plots and results.*

9. Compute some summary statistics on the provided data.
Scroll down to the 'Basic summary for the counts' heading. 
Here you will see the total number of uniquely assigned reads per sample, accompanied by some information about the expressed genes in the data, specified via simple thresholds. 
The first option is provided through setting the 'Threshold on the row sums of counts' option. 
To test this option, enter 10 in the dedicated field and compare the results to the ones shown in Fig. X. 
The second option to analyze the genes in the data is the 'Threshold on the row means of the normalized counts' option (generally stricter than the former definition). 
Enter 10 for this option and compare the results with the ones shown in Fig. X. 

    **TODO: Something about how evaluating the number of counts per gene in your sample can help with your analysis **

![**Figure X** Basic summary of the counts - The Data Overview panel provides different summaries on the number of counts in the samples. Among these is the number of uniquely aligned reads in each sample and the number of genes with non-zero counts in the data.](Figures_Basic_1/Figure 10.png)

**Explore the Principal Components from the samples point of view.**

10. Create a Principal Component plot.
Navigate to the Samples View panel (Fig. X). 
In this panel you will see the PCA of the individual samples of the input data on the left side, with the corresponding scree plot shown on the right side. 
If any value is selected in the 'Group/color by' field, users can explore the impact of each covariate by means of the color encoding applied to the graphics.
Use the 'Group/color by' option to color the samples by condition - the PCA plot should now display the samples in 4 colors (Fig. X). 
You can also download the PCA plot via the 'Download plot' button in the lower right corner; as seen before, you can specify the name of the file you want to save the plot. 
This is also possible for the scree plot on the right side of the panel. 

    *The PCA plot of the sample data indicates that the most variance in the data seems to arise from the condition of the individual samples in this data set. 
    When analysing data, it is important for users to know which characteristic of their samples has the most influence (i.e. is responsible for the most variance). 
    Hence, users should select different options from the 'Group/color by' option to discover the characteristic with the most influence. 
    The scree plot right next to the PCA can also help users to evaluate how many principal components should be used to analyse the variance in the data. 
    In the example shown in this protocol, it can be easily observed that the first two components explain the nearly 80% of the variance in the data. 
    80% can be seen as a good cut-off value to select the number of principal components.*

  
![**Figure ** The Samples View panel - This panel shows a PCA of the individual samples of the input data. In the upper part of the panel, the PCA on the samples and the corresponding scree plot are shown. Initially there is no grouping shown in the PCA.](Figures_Basic_1/Figure 11.png)

![**Figure ** PCA colored by condition - The PCA plot of the samples is shown colored by condition. This coloring is achieved by selecting the option 'condition' in the 'Group/color by' selection tab at the far left of the screen. The tab indicates the selected option by displaying the name of the option in the field. Users can remove options by clicking on the option field right behind the option to be deleted and hitting the delete key on the keyboard.](Figures_Basic_1/Figure 12.png)

11. Change the parameters to display different versions of the PCA.
Navigate to the options panel on the far left and search for the 'Nr. of (most variant) genes' option, set as default to 300 genes. 
Change this number to 500 - either using the increase and decrease buttons in the option field (indicated by small arrows) or typing the value in the box.
Inspect the PCA plot and compare it with the plot in Fig. X. 
The values of the percent of variance explained of the first and second principal component should have changed, and according to the values, you might observe slight differences in the placement and clustering of the samples (according to how the inclusion of a specific set of genes influences the overall variability).

    *The PCA plot visualizes the variance in the data and the differences between the individual samples. 
    The number of genes selected for the PCA plot can influence the overall structure of the plot. 
    In order to analyze which characteristic is responsible for the variance in the data, users should also consider to change the number of genes used for the PCA. 
    This enables users to analyze the robustness of the identified characteristic.*
    
Users can also change the principal components that are displayed on either axis: by default, PC1 and PC2 are shown, but this can be changed by the select widgets in the left sidebar.
This can be extremely relevant when analyzing large datasets, where it is hard to identify a clean structure - in this Basic Protocol, the dimensions of variability are relatively straightforward to define and associate to experimental covariates.

12. Zoom in the PCA plot.
The Sample View panel provides the possibility to zoom in on the PCA, using the Shiny brushing mechanism: hover your mouse over the PCA plot until your mouse cursor changes to a little plus sign. 
Click and hold the left mouse button to draw a little grey rectangle in the main plot (Fig. X).
Move the mouse to the upper left corner of the PCA plot and select all samples belonging to the naive condition (colored in blue in Fig. X) as explained before. 
Below the main plot, you can find a zoomed version of the PCA plot, containing mainly the selected samples. 
This should somewhat look like the plot shown in Fig. X. 

    *The zoom functionality is especially helpful in cluttered PCA plots with many samples or particular cluttered regions of the plot.*


![**Figure ** Zoomed PCA plot - The figure shows the zooming mechanism on the PCA plot of the Samples View panel. The grey rectangle in the upper PCA plot indicates the section selected for zooming. This selected section of the plot is shown in the lower PCA plot in more detail.](Figures_Basic_1/Figure 13.png)

13. Identify genes with high loadings on the selected principal components.
As a result to the computation of the principal components, the genes have a specific value for the loadings, reflecting the "weight" that each of them has on a particular principal component - this can be thought of as the correlation between the PC and the original variable.
For the two selected principal components (by default, PC1 and PC2), the genes with highest loadings (in either direction along the axis of a PC) are displayed as bar chart, and can later be followed up, e.g. in the Gene Finder tab panel.

14. Inspect the impact of marking samples as outliers.
Scroll down in the Sample View panel until you see the 'Outlier identification' option (Fig. X). 
The 'Select which sample(s) to remove - suspected outliers' option provides the possibility to select and remove individual samples from the PCA. 
For demonstration purposes select all samples of the 'naive' condition and remove them from the PCA (the sample ids end in '102', '111', '182', '262', '284', and '368'). 
The options included in the drop-down are the samples still used for the calculation. 
Select the above mentioned sample ids until the PCA plot looks like Fig. X.

    *The sample outlier detection option can help users at identifying individual samples which are highly different from the remaining samples and hence could be possible outliers. 
    Especially for data sets with small sample numbers, the outlier identification option is useful to evaluate the influence of individual samples on the overall PCA of the data. 
    At the bottom of the Sample View panel, users also have the option to select a third principal component to visualize besides the first two. 
    This is especially helpful in data sets where the most variance of the data is scattered among several principal component and not only the first two.*
  
  **TODO: Maybe we can merge the next two figures**

![**Figure ** Outlier identification - The Samples View panel provides the possibility to detect outliers.](Figures_Basic_1/Figure 14.png)

![**Figure ** PCA plot with outliers removed - The figure shows the PCA plot with all samples of the 'naive' condition removed. With this type of sample selection, outlier samples in the data can be detected and removed from the PCA.](Figures_Basic_1/Figure 15.png)

**Explore the Principal Components from the genes point of view.**

15. Create the genes biplot.
Navigate to the Genes View panel (Fig. X), where you will see a PCA on the individual genes as a dual view to samples PCA from the previous steps. 
Again, you can select a section of the plot to zoom which will be shown on the right side in the 'Zoomed window' (Fig. X). 
Upon selecting a subset of genes in the main plot, these will also be displayed in the profile explorer as scaled expression values - a value needs to be selected in the 'Group/color by' selection, and you can continue using the 'condition' as in the previous steps (Fig. X). 

    *(**TODO: explain what the profile explorer plot shows**)* 
    
    **TODO: Maybe we can also shown only one of the following pictures and mentioning the zoom functionality.**

![**Figure ** Genes View panel - The Genes View panel provides a PCA plot of the individual genes as well as the possibility to select section for a zoomed view.](Figures_Basic_1/Figure 16.png)
![**Figure ** Zoomed in PCA plot - The zoomed view of the PCA plot visualizes the genes highlighted by the grey rectangle in the left PCA plot.](Figures_Basic_1/Figure 17.png)

16. Inspect single selected genes.
In the 'Zoomed window', select a gene by clicking on one of the black dots. 
This will generate a box plot under the 'Boxplot of selected gene' heading (Fig. X), showing the normalized counts for the selected gene, with the conditions defined by the value in the 'Group/color by' selector. 
You can also change the style of the plot to a violin plot. To change the plot, search for the 'Plot style for gene counts' option in the option bar on the left. 


![**Figure** Profile Explorer and Box plot - ](Figures_Basic_1/Figure 18.png)

17. Inspect subsets of selected genes.
Below the 'Profile explorer' and 'Boxplot of selected gene' plot you will find two versions of the heatmap for the selected genes (displayed in the zoomed window) in all the individual samples (Fig. X) - one static (left) and one interactive, based on the Plotly framework (right). 
On the bottom, the collapsible element 'Table export options' contains the tabular information for the same subset, and offers the possibility to export their content.

![**Figure ** Heatmap Gene View panel - ](Figures_Basic_1/Figure 19.png)

18. Inspect genes of interest in the GeneFinder.
Navigate to the GeneFinder panel (Fig. X, **TODO: Grafik erstellen**). 
Enter 'TSPAN6' in the 'Type in the name of the gene to search' option field.
Make sure that 'condition' is still selected in the 'Group/color by' option; otherwise, select it from the available option.
The panel will load a box plot of the normalized counts of the selected gene as well as a table with the normalized counts of the gene for each sample.
The individual sample names are annotated on the plotted points, and the underlying data is also reported as a compact table.

    *The Gene Finder panel can be helpful in identifying the counts of a specific gene of interest in the individual samples. 
    If the gene of interest is not included in the data, the panel will inform the user by displaying the error message 'Could not find the gene you typed'. 
    The panel will also inform users if the gene name is misspelled and will suggest a gene name close to the typed one. 
    The search functionality is case sensitive and has no autocomplete option, so it is important to type the gene name correctly. *

19. Generate a functional interpretation for the principal components.
Navigate to the PCA2GO panel (Fig. X, **TODO: Grafik erstellen**). 
Click on the 'Select species for your sample' option, and from the drop-down menu select 'human'. 
The 'Select the input type of your identifiers' option can remain unaltered, as we are already using ENSEMBL identifiers; if this is not the case in your own data, make sure to check what identifiers are in use, and select them accordingly. 
Click the blue 'Compute the PCA2GO object' button, the application will compute for each PC and for each direction a set of enrichment results - it might take a while to compute the object.
Each enrichment table will be displayed on either side of the PCA plot in the middle of the panel, and can be used to pinpoint biological processes and functions that can explain the observed variability on the latent space representation
Optionally, to save time and obtain more fine-grained functional categories, users can compute such an object beforehand with methods implemented in the topGO package, which are conveniently wrapped by the `pca2go` function.
We refer to the package vignette where its specific usage is fully documented.

**Wrapping up the analysis with pcaExplorer.**

20. Generate, preview, and export an analysis report.
Navigate to the Report Editor panel (Fig. X, **TODO: Grafik erstellen**). 
Open the 'Markdown options' drop-down menu, and enter as a title for the report 'Current Protocols pcaExplorer report' and your name in the 'Author' field; you can leave all other options unaltered. 
Click the 'Update report' button, and scroll down in the panel to see an preview of the HTML report in the web application itself. 
Click on the 'Generate & Save' button next to the 'Update report' button to download the report. 
Save the report under a chosen name on your computer - this operation might take a while, as most content is generated from a fresh session, based on the current values for the reactive elements in the app.

21. Find additional information on pcaExplorer.
Navigate to the Instructions panel, where two buttons can be clicked to open and inspect the pcaExplorer vignettes (available when installing the package)  (Fig. X **TODO: Grafik erstellen**). 
These vignettes describe comprehensively the pcaExplorer package, and constitute an excellent complement to this Basic Protocol when searching for additional, up-to-date documentation.
A quickstart guide is also displayed if clicking on the "Up and running with pcaExplorer" collapsible element.
If users are interested in additional information about the package, the About panel (Fig. X **TODO: Grafik erstellen**) lists also the developers contact information and the a citation entry.
For the sake of computational reproducibility, consider adding the output of the sessionInfo section into your electronic notebook documentation, to simplify the compilation of the "Materials and Methods" section for a manuscript where you used pcaExplorer.

![**Figure ** Instructions panel](Figures_Basic_1/Figure last.png)


# Session information {-}

```{r}
sessionInfo()
```

# References {-}
